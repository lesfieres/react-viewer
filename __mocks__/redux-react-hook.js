/* globals jest:true */
export const useMappedState = jest.fn(() => {});
export const useDispatch = jest.fn(() => () => {});
