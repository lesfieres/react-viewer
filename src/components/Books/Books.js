import React, { useState, useCallback, useEffect } from 'react'; // eslint-disable-line no-unused-vars, prop-types
import { useDebounce } from 'react-use';
import { useMappedState, useDispatch } from 'redux-react-hook';
import TextField from '@material-ui/core/TextField';
import { withStyles } from '@material-ui/core';
import InputAdornment from '@material-ui/core/InputAdornment';
import SearchIcon from '@material-ui/icons/Search';
import CircularProgress from '@material-ui/core/CircularProgress';
import Grid from '@material-ui/core/Grid';
import Item from '../Item/Item';
import { fetchBooks } from '../../redux/actions/bookActions';
import ItemDisplaySelect, {
  DisplayCardType,
} from '../ItemDisplaySelect/ItemDisplaySelect';

const styles = {
  searchInput: {
    color: '#fff',
  },
};

export const BooksNoStyled = props => {
  const { classes } = props;
  const [inputSearch, setInputSearch] = useState('');
  const [performingSearch, setPerformingSearch] = useState(false);
  const [currentDisplayCardType, setCurrentDisplayCardType] = useState(
    DisplayCardType.BigCard,
  );

  const mapState = state => state.books;
  const { books } = useMappedState(mapState);

  const dispatch = useDispatch();
  const fetchBooksCallback = useCallback(fetchBooks(inputSearch, dispatch));
  useDebounce(fetchBooksCallback, 500, [inputSearch]);

  useEffect(() => {
    setPerformingSearch(true);
  }, [inputSearch]);

  useEffect(() => {
    setPerformingSearch(false);
  }, [books]);

  const createGridItem = item => {
    switch (currentDisplayCardType) {
      case DisplayCardType.List:
        return (
          <Grid item key={item.id} xs={12}>
            <Item item={item} />
          </Grid>
        );
      case DisplayCardType.SmallCard:
        return (
          <Grid item key={item.id} xs={12} md={4} lg={2}>
            <Item item={item} />
          </Grid>
        );
      case DisplayCardType.BigCard:
      default:
        return (
          <Grid item key={item.id} xs={12} md={6} lg={4}>
            <Item item={item} />
          </Grid>
        );
    }
  };

  return (
    <React.Fragment>
      <TextField
        id="standard-search"
        label="Search"
        type="search"
        className={classes.searchInput}
        margin="normal"
        fullWidth
        value={inputSearch}
        onChange={e => setInputSearch(e.target.value)}
        InputProps={{
          endAdornment: (
            <InputAdornment position="end">
              <SearchIcon />
            </InputAdornment>
          ),
        }}
      />
      <ItemDisplaySelect
        displayType={currentDisplayCardType}
        setCurrentDisplayCardType={setCurrentDisplayCardType}
      />
      {performingSearch ? (
        <CircularProgress className={classes.progress} />
      ) : (
        <Grid container spacing={24} className="item-list">
          {!performingSearch && books.map(createGridItem)}
        </Grid>
      )}
    </React.Fragment>
  );
};

const Books = withStyles(styles)(BooksNoStyled);
export default Books;
