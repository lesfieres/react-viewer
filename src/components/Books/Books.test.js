/* eslint-disable prop-types */
// eslint-disable-next-line no-unused-vars
import React from 'react';
import { useMappedState } from 'redux-react-hook';
import ShallowRenderer from 'react-test-renderer/shallow';
import Books from './Books';

describe('Books component test', () => {
  const renderer = new ShallowRenderer();

  it('renders without items', () => {
    useMappedState.mockImplementation(() => ({
      books: [],
    }));

    renderer.render(<Books classes={{}} />);
    const result = renderer.getRenderOutput();

    expect(result).toMatchSnapshot();
  });

  it('renders with items', done => {
    useMappedState.mockImplementation(() => ({
      books: [{ id: '1', author: {} }, { id: '2', author: {} }],
    }));

    renderer.render(<Books classes={{}} />);
    const result = renderer.getRenderOutput();

    setTimeout(() => {
      expect(result).toMatchSnapshot();
      done();
    }, 1000);
  });
});
