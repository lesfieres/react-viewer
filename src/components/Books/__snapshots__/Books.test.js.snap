// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Books component test renders with items 1`] = `
<React.Fragment>
  <TextField
    InputProps={
      Object {
        "endAdornment": <WithStyles(WithFormControlContext(InputAdornment))
          position="end"
        >
          <pure(SearchIcon) />
        </WithStyles(WithFormControlContext(InputAdornment))>,
      }
    }
    fullWidth={true}
    id="standard-search"
    label="Search"
    margin="normal"
    onChange={[Function]}
    required={false}
    select={false}
    type="search"
    value=""
    variant="standard"
  />
  <ItemDisplaySelect
    displayType="DISPLAY_CARD_TYPE_BIG_CARD"
    setCurrentDisplayCardType={[Function]}
  />
  <WithStyles(Grid)
    className="item-list"
    container={true}
    spacing={24}
  >
    <WithStyles(Grid)
      item={true}
      lg={4}
      md={6}
      xs={12}
    >
      <Item
        item={
          Object {
            "author": Object {},
            "id": "1",
          }
        }
      />
    </WithStyles(Grid)>
    <WithStyles(Grid)
      item={true}
      lg={4}
      md={6}
      xs={12}
    >
      <Item
        item={
          Object {
            "author": Object {},
            "id": "2",
          }
        }
      />
    </WithStyles(Grid)>
  </WithStyles(Grid)>
</React.Fragment>
`;

exports[`Books component test renders without items 1`] = `
<React.Fragment>
  <TextField
    InputProps={
      Object {
        "endAdornment": <WithStyles(WithFormControlContext(InputAdornment))
          position="end"
        >
          <pure(SearchIcon) />
        </WithStyles(WithFormControlContext(InputAdornment))>,
      }
    }
    fullWidth={true}
    id="standard-search"
    label="Search"
    margin="normal"
    onChange={[Function]}
    required={false}
    select={false}
    type="search"
    value=""
    variant="standard"
  />
  <ItemDisplaySelect
    displayType="DISPLAY_CARD_TYPE_BIG_CARD"
    setCurrentDisplayCardType={[Function]}
  />
  <WithStyles(Grid)
    className="item-list"
    container={true}
    spacing={24}
  />
</React.Fragment>
`;
